### ISSUE TITLE
Change Request: Make processors (pre/post) async or pass config to them

### ISSUE THREAD

**Issue by eladavron** [eslint/eslint#15494]
### ESLint version

8.6.0

### What problem do you want to solve?

ESLint currently  calls the preprocessor and postprocessor functions synchronously.
This is a problem since the ESLint class is asynchronous, thus it's currently impossible to use it within the preprocessor.

For example, my use case is I need to know during preprocessing what type of quotes the linter will be using.  

With CLIEngine it was just:
```js
preprocess(text, filename) {
    const config = new CLIEngine({ useEslintrc: true }).getConfigForFile(path.resolve(filename));
    const quote = config.rules.quotes[1];
   /* ... */
}
```

However with ESLint, the equivalent function is `getConfigForFile` which is async.
If you try to just make the processor async, like so:
```js
async preprocess(text, filename) {
    const eslint = new ESLint();
    const config = await eslint.calculateConfigForFile(filename);
    /* ... */
}
```
... then it is now returning a `Future` instead of an array which eslint doesn't know how to process.

### What do you think is the correct solution?

Either make the the processors async as well and have eslint treat them as such, or pass the loaded config to the processor, like:

```js
preprocess(text, filename, config) {
    const quote = config.rules.quotes;
   /* ... */
}
```

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
Hmm, this is really not a use case that ESLint is designed for. Processors are not intended to enforce conventions, just to transform input into a format that ESLint can understand. Can you explain more about why you want to know which quotes ESLint is enforcing inside a processor?

Also, Where is this code being used? Is it a public utility?

Some feedback on your suggestions:

1. we can’t make processors async because they run inside of Linter, which is synchronous.
2. We can’t pass a config to processors because config is calculated and the config used for any given snippet returned from a processor might be different than the one we have before the processor is called.

**Comment by eladavron** [comment]
Yeah that's the thing.  
What I'm trying to do is revive this old plugin: https://github.com/alexkuz/eslint-plugin-jinja
It transforms Jinja syntax into regular JS syntax that ESLint can understand.  
Specifically the use for quotes is converting something like this:
```js
var foo = {{some_value}}
```
to this:
```js
var foo = "{{some_value}}"
```
while retaining the correct quotes usage.
If you don't use the quotes defined in the rules, it just gives you a "wrong quotes" error on that.

**Comment by nzakas** [comment]
Yeah, this isn’t a good case for ESLint to solve. It seems like a better approach to use a separate file of some sort to specify preferred quotes and read from that file in the processor rather than trying to pull ESLint config. Maybe you could even use an editorconfig file?

**Comment by eladavron** [comment]
Thanks.
It just occurred to me though, I can replace it with a number instead of a string - it will have effectively the same purpose as a string but I won't have to worry about quotes.
I just have to make sure it has the same number of digits as the original expression does to avoid having errors appear on the wrong columns.

**Comment by nzakas** [comment]
Sounds like a good approach. 👍
